pronto_mia/lib/core/services/push_notification_service.dart

```diff
@@ -73,14 +88,89 @@ class PushNotificationService {
     }
   }
 
-  // TODO: Improve foreground message handling
-  Future<void> _handleForegroundMessage(RemoteMessage message) async {
-    (await _loggingService)
-        .log("PushNotificationService", Level.INFO, "Message was received!");
+  Future<void> handleInitialMessage() async {
+    final initialMessage = await _fcm.getInitialMessage();
+    if (initialMessage != null) {
+      _handleMessage(initialMessage);
+    }
+  }
+
+  Future<void> _handleMessage(RemoteMessage message) async {
+    if (message.data != null) {
+      final action = message.data['Action'].toString();
+      switch (action) {
+        case "publish":
+          (await _loggingService).log(
+              "PushNotificationService", Level.INFO, 'Message is a publish');
+          await _handlePublishMessages(message);
+          break;
+        default:
+          (await _loggingService).log("PushNotificationService", Level.WARNING,
+              'Message contained an unknown action: $action');
+      }
+    }
+  }
+
+  Future<void> _handlePublishMessages(RemoteMessage message) async {
+    final targetType = message.data['TargetType'].toString();
+    switch (targetType) {
+      case "deploymentPlan":
+        (await _loggingService).log("PushNotificationService", Level.INFO,
+            'Message is a deploymentPlan');
+        await _handleDeploymentPlanPublish(message);
+        break;
+      default:
+        (await _loggingService).log("PushNotificationService", Level.WARNING,
+            'Message contained an unknown target: $targetType');
+    }
+  }
+
+  Future<void> _handleDeploymentPlanPublish(RemoteMessage message) async {
+    final targetId = int.parse(message.data['TargetId'].toString());
+    final deploymentPlan =
+        await _deploymentPlanService.getDeploymentPlan(targetId);
 
-    if (message.notification != null) {
-      (await _loggingService).log("PushNotificationService", Level.INFO,
-          'Message contained a notification: ${message.notification.body}');
+    if (deploymentPlan == null) {
+      return;
     }
+
+    final dateFormat = DateFormat('dd.MM.yyyy');
+    final availableFromFormatted =
+        dateFormat.format(deploymentPlan.availableFrom);
+    final availableUntilFormatted =
+        dateFormat.format(deploymentPlan.availableUntil);
+    final pdfViewArguments = PdfViewArguments(
+      pdfPath: deploymentPlan.link,
+      title: deploymentPlan.description ?? 'Einsatzplan',
+      subTitle: '$availableFromFormatted - $availableUntilFormatted',
+    );
+
+    final dialogResponse = await _dialogService.showCustomDialog(
+        variant: DialogType.custom,
+        customData: Container(

```

Comment:
I suggest you extract the dialog content into its own stateless widget, so the gui specific code is not in the service itself.